### ISSUE TITLE
Homebrew installed but Cakebrew not detecting it

### ISSUE THREAD

**Issue by vinyldarkscratch** [brunophilipe/Cakebrew#202]
Unfortunately, Cakebrew doesn't seem to be detecting my Homebrew installation.  I don't have any special install tricks or fancy paths; Homebrew is installed the standard way.  Objective-C isn't my forté, sadly, may I ask how you're detecting if Homebrew is installed?

**Comment by iStyx** [comment]
Same thing here.

```sh
$ brew config
HOMEBREW_VERSION: 1.5.12-40-gf37b0bc
ORIGIN: https://github.com/Homebrew/brew
HEAD: f37b0bc9876e196b8b7b7329c7a6572303f5c855
Last commit: 2 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 5b66a88e2f886871e0044c42b7453e739b739f8e
Core tap last commit: 6 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_DEV_CMD_RUN: 1
HOMEBREW_GITHUB_API_TOKEN: set
CPU: octa-core 64-bit sandybridge
Homebrew Ruby: 2.3.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3/bin/ruby
Clang: 9.0 build 900
Git: 2.14.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.54.0 => /usr/bin/curl
Perl: /usr/bin/perl
Python: /usr/bin/python
Ruby: /usr/bin/ruby => /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/ruby
Java: 9
macOS: 10.13.3-x86_64
Xcode: 9.2
CLT: N/A
X11: N/A
```

```sh
$ brew cask --version
Homebrew-Cask 1.5.12-40-gf37b0bc
caskroom/homebrew-cask (git revision 03f7f; last commit 2018-03-25)
```

```sh
$ brew cask info cakebrew
cakebrew: 1.2.5
https://www.cakebrew.com/
/usr/local/Caskroom/cakebrew/1.2.5 (64B)
From: https://github.com/caskroom/homebrew-cask/blob/master/Casks/cakebrew.rb
==> Name
Cakebrew
==> Artifacts
Cakebrew.app (App)
```

**Comment by iStyx** [comment]
Seems like the source of this issue is my using `tcsh` shell. Why not just invoke `bash` instead of user's default shell?

**Comment by vinyldarkscratch** [comment]
That would make sense, I'm using `zsh` as my default shell.  So, this is a problem with using other shell prompts, it seems!

**Comment by brunophilipe** [comment]
ZSH behaves differently in login mode, have a look at this:

https://github.com/brunophilipe/Cakebrew/wiki#using-zsh-shell

**Comment by iStyx** [comment]
@brunophilipe If `csh/tcsh` are not supported at all, why not just invoke `bash` if user's shell is one of those?

**Comment by brunophilipe** [comment]
Because if the user is not using bash, than the PATH variable is probably not set either. I suppose Cakebrew could try it, and at the same time I’m afraid this could cause even more confusion :/

**Comment by vinyldarkscratch** [comment]
No luck, @brunophilipe -- tried setting my `PATH` in `~/.zlogin` and `~/.zprofile` (both I had to create myself), and it's already set in `~/.zshrc`.

**Comment by brunophilipe** [comment]
@vinyldarkscratch what’s the output of `bash -e -c "echo $SHELL"`? The output should be the same if ran from any terminal, unless the value of `$SHELL` is being overridden somewhere.

**Comment by iStyx** [comment]
@brunophilipe I'm not using `bash` at all, although when I invoke `bash` shell (*not* from `tcsh` shell), my `$PATH` is:
```sh
$ echo $PATH
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/MacGPG2/bin
```
That's because of content of `/etc/paths` and `/etc/paths.d/*`. I can certainly say I *did not* add `/usr/local/bin` there.

**Comment by vinyldarkscratch** [comment]
@brunophilipe As expected, it's `/bin/sh`.  Checking my `PATH` variable in both `zsh` and `bash`, they both have `/usr/local/bin` in them, which is where Brew is installed.

**Comment by brunophilipe** [comment]
I was re-reading the code, and I realized Cakebrew actually uses `NSProcessInfo` to find out the user shell. According to the documentation, -[NSProcessInfo environment]` returns "The variable names (keys) and their values in the environment from which the process was launched."

How are you launching Cakebrew?

**Comment by EricRFMA-ES** [comment]
Having the same issue.  `brew` is installed in `/usr/local/bin`, which is on my `$PATH`. I _am_ using `tcsh`, but if I invoke `bash` (or `sh`) `-i`, `/usr/local/bin` is on the path. In fact, even though I use `tcsh`, I have `/usr/local/bin` added to the path in my `.bashrc`.

I've even gone as far as invoking `bash -i` then going to `/Applications/Cakebrew/Contents/MacOS` and doing a `open -a Cakebrew`.

In no case can `Cakebrew` find my `brew` installation.

It goes without saying (but I will anyway) that double clicking on `Cakebrew.app` results in the brew not found dialog coming up.

**Comment by iStyx** [comment]
@brunophilipe It's hard to understand to whom you're asking a question since you do not mention any username :)

**Comment by EricRFMA-ES** [comment]
My Objective-C isn't very strong, but from looking at the code I found the following:

The code tries to run what amounts to: `$SHELL -l -c "which brew"` where `$SHELL` is your login shell. For `tcsh` (and friends) this is a problem because they don't have a `-l` flag, so the command fails.

I'd suggest explicitly using `/bin/bash` instead of `$SHELL`. I don't know if _every_ version of MacOS you support has `/bin/bash` but I think they do. It just seems a better practice to specifically use a known shell with known parameters instead of just grabbing the login shell which is an unknown quantity and hoping for the best.

I hope this helps.

**Comment by iStyx** [comment]
`tcsh` does have `-l` flag, but it is applicable only in case it is the only flag specified.
Actually, I don't think this `-l` is needed at all. macOS does have `/usr/local/bin` in its paths by default.
```sh
$ bash -c "which brew"
/usr/local/bin/brew
$ zsh -c "which brew"
/usr/local/bin/brew
$ tcsh -c "which brew"
/usr/local/bin/brew
$ csh -c "which brew"
/usr/local/bin/brew
$ sh -c "which brew"
/usr/local/bin/brew
```

**Comment by darkn3rd** [comment]
I am having this issue as well.  I expect the program to be able to detect that brew is installed.  I don't care the mechanism in which it users, as an end user, I expect that if brew is installed, then CakeBrew would work, and not block itself from functioning.  That being said, as a developer, checking some shell setup is maybe not the best way to go about it, because then you have to have some advance patterns to check all the potential outcomes, otherwise, you have the Blame the User scenario.

Bottom line, program blocks itself from operating, cannot reasonably detect that Homebrew is properly installed.  I don't see the problem with accepting the issue?  Are we disputing that Homebrew is not installed and the program detected that brew was not really brew?

**Comment by brunophilipe** [comment]
If you have nothing to add to this issue than I would request you to stop posting useless messages to this case, which are mailed to all the users subscribed to it.

Please refer to the Cakebrew license where it states that no warranty of any kind is provided.

I do not have time to improve this feature right now, and it seems you don’t have code to contribute either.

**Comment by vinyldarkscratch** [comment]
Well, here's a repository I'm not going to be using if/once this issue is fixed.  :/

**Comment by PC-Chiu** [comment]
`~ /bin/bash -l -c "which brew"`
/opt/homebrew/bin/brew

`sudo ln -s /opt/homebrew/bin/brew /usr/local/bin/brew`
